docs/unittesting.md

```diff
@@ -0,0 +1,264 @@
+# Unit Testing in TorQ
+
+### Introduction
+
+Unit testing is a software validation methodology where individual modules or 'units' of source code are tested to determine whether they work as intended. TorQ's unit testing framework has been designed to be user-friendly and includes lots of useful features such as debugging and integration testing. We shall see that, using this framework, it is straightforward to test both small units of code as well as larger pieces of application functionality.
+
+### Test Basics
+
+**My First Test**
+
+Tests are written in the style of Kx's [k4unit](https://code.kx.com/q/kb/unit-tests/), where tests are written in CSV files and loaded into a Q process along with the testing code and are run line by line, with the results being stored in a table. In order to write and run a basic test, in a new folder create a CSV file with some basic tests in it, like the following:
+
+```
+action,ms,bytes,lang,code,repeat,minver,comment
+comment,0,0,,this will be ignored,1,,""
+before,0,0,q,aa:22,1,,"This sets a variable before the tests begin"
+true,0,0,q,2=sum 1 1,1,,"A basic arithmetic check"
+fail,0,0,q,2=`aa,1,,"This ought to fail"
+after,0,0,q,bb:33,1,,"This code executes after the tests have run"
+```
+
+Then use the `torq.q` script to start up a TorQ process with the test flag pointing to the directory where the CSV is written and a debug flag so that the process outputs its logs into the prompt and stays alive:
+
+```shell
+$ q ${TORQHOME}/torq.q -proctype test -procname test1 -test /path/to/my/tests -debug
+```
+
+This will automatically load in the test code and all CSVs in that directory into a test TorQ process and run the tests in those files, then output the results to the screen.
+
+**Simple RDB Test**
+
+A similar scenario is when the test code is loaded into a 'production' TorQ process, ie. an RDB or a tickerplant, in order to test the functioning of that process. A simple RDB example can be seen here:
+
+```
+action,ms,bytes,lang,code,repeat,minver,comment
+true,0,0,q,`segmentedtickerplant~.rdb.tickerplanttypes,1,,"Check the TP type"
+true,0,0,q,all (null;{-11h=type x}) @' .rdb.subscribeto,1,,"Check RDB is subscribed to null symbol (all tables)"
```

Comment:
think this is the same? unless there's a reason to make this more complex

```suggestion
true,0,0,q,`~.rdb.subscribeto,1,,"Check RDB is subscribed to null symbol (all tables)"
```

FWIW, I'm not sure that these are great example tests, as it's just checking the value of some configurable variables, rather than actually testing any functionality? I suppose the idea is to keep it simple for example, but IMO it'd be better to show testing actual funcationality